### PR TITLE
Create PFB-based verbatim manifest format for HCA (#6122)

### LINKED ISSUE #6122: Create PFB-based verbatim manifest format for HCA

### PR DESCRIPTION
**Pull request by nadove-ucsc**
Connected issues: #6122

### Author

- [x] PR is a draft
- [x] Target branch is `develop`
- [x] Name of PR branch matches `issues/<GitHub handle of author>/<issue#>-<slug>`
- [x] On ZenHub, PR is connected to all issues it (partially) resolves
- [x] PR description links to connected issues
- [x] PR title matches<sup>1</sup> that of a connected issue <sub>or comment in PR explains why they're different</sub>
- [x] PR title references all connected issues
- [x] For each connected issue, there is at least one commit whose title references that issue

<sup>1</sup> when the issue title describes a problem, the corresponding PR
title is `Fix: ` followed by the issue title

### Author (partiality)

- [x] Added `p` tag to titles of partial commits
- [x] This PR is labeled `partial` <sub>or completely resolves all connected issues</sub>
- [x] This PR partially resolves each of the connected issues <sub>or does not have the `partial` label</sub>

### Author (chains)

- [x] This PR is blocked by previous PR in the chain <sub>or is not chained to another PR</sub>
- [x] The blocking PR is labeled `base` <sub>or this PR is not chained to another PR</sub>
- [x] This PR is labeled `chained` <sub>or is not chained to another PR</sub>

### Author (reindex, API changes)

- [x] Added `r` tag to commit title <sub>or the changes introduced by this PR will not require reindexing of any deployment</sub>
- [x] This PR is labeled `reindex:dev` <sub>or the changes introduced by it will not require reindexing of `dev`</sub>
- [x] This PR is labeled `reindex:anvildev` <sub>or the changes introduced by it will not require reindexing of `anvildev`</sub>
- [x] This PR is labeled `reindex:anvilprod` <sub>or the changes introduced by it will not require reindexing of `anvilprod`</sub>
- [x] This PR is labeled `reindex:prod` <sub>or the changes introduced by it will not require reindexing of `prod`</sub>
- [x] This PR is labeled `reindex:partial` and its description documents the specific reindexing procedure for `dev`, `anvildev`, `anvilprod` and `prod` <sub>or requires a full reindex or carries none of the labels `reindex:dev`, `reindex:anvildev`, `reindex:anvilprod` and `reindex:prod`</sub>
- [x] This PR and its connected issues are labeled `API` <sub>or this PR does not modify a REST API</sub>
- [x] Added `a` (`A`) tag to commit title for backwards (in)compatible changes <sub>or this PR does not modify a REST API</sub>
- [x] Updated REST API version number in `app.py` <sub>or this PR does not modify a REST API</sub>

### Author (upgrading deployments)

- [x] Ran `make image_manifests.json` and committed the resulting changes <sub>or this PR does not modify `azul_docker_images`, or any other variables referenced in the definition of that variable</sub>
- [x] Documented upgrading of deployments in UPGRADING.rst <sub>or this PR does not require upgrading deployments</sub>
- [x] Added `u` tag to commit title <sub>or this PR does not require upgrading deployments</sub>
- [x] This PR is labeled `upgrade` <sub>or does not require upgrading deployments</sub>
- [x] This PR is labeled `deploy:shared` <sub>or does not modify `image_manifests.json`, and does not require deploying the `shared` component for any other reason</sub>
- [x] This PR is labeled `deploy:gitlab` <sub>or does not require deploying the `gitlab` component</sub>
- [x] This PR is labeled `deploy:runner` <sub>or does not require deploying the `runner` image</sub>

### Author (hotfixes)

- [x] Added `F` tag to main commit title <sub>or this PR does not include permanent fix for a temporary hotfix</sub>
- [x] Reverted the temporary hotfixes for any connected issues <sub>or the none of the stable branches (`anvilprod` and `prod`) have temporary hotfixes for any of the issues connected to this PR</sub>

### Author (before every review)

- [x] Rebased PR branch on `develop`, squashed old fixups
- [x] Ran `make requirements_update` <sub>or this PR does not modify `requirements*.txt`, `common.mk`, `Makefile` and `Dockerfile`</sub>
- [x] Added `R` tag to commit title <sub>or this PR does not modify `requirements*.txt`</sub>
- [x] This PR is labeled `reqs` <sub>or does not modify `requirements*.txt`</sub>
- [x] `make integration_test` passes in personal deployment <sub>or this PR does not modify functionality that could affect the IT outcome</sub>

### Peer reviewer (after approval)

- [x] PR is not a draft
- [x] Ticket is in *Review requested* column
- [x] PR is awaiting requested review from system administrator
- [x] PR is assigned to only the system administrator

### System administrator (after approval)

- [x] Actually approved the PR
- [x] Labeled connected issues as `demo` or `no demo`
- [x] Commented on connected issues about demo expectations <sub>or all connected issues are labeled `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] A comment to this PR details the completed security design review
- [x] PR title is appropriate as title of merge commit
- [x] `N reviews` label is accurate
- [x] Moved ticket to *Approved* column
- [x] PR is assigned to only the operator

### Operator (before pushing merge the commit)

- [x] Checked `reindex:…` labels and `r` commit title tag
- [x] Checked that demo expectations are clear <sub>or all connected issues are labeled `no demo`</sub>
- [x] Squashed PR branch and rebased onto `develop`
- [x] Sanity-checked history
- [x] Pushed PR branch to GitHub
- [x] Ran `_select dev.shared && CI_COMMIT_REF_NAME=develop make -C terraform/shared apply_keep_unused` <sub>or this PR is not labeled `deploy:shared`</sub>
- [x] Ran `_select dev.gitlab && CI_COMMIT_REF_NAME=develop make -C terraform/gitlab apply` <sub>or this PR is not labeled `deploy:gitlab`</sub>
- [x] Ran `_select anvildev.shared && CI_COMMIT_REF_NAME=develop make -C terraform/shared apply_keep_unused` <sub>or this PR is not labeled `deploy:shared`</sub>
- [x] Ran `_select anvildev.gitlab && CI_COMMIT_REF_NAME=develop make -C terraform/gitlab apply` <sub>or this PR is not labeled `deploy:gitlab`</sub>
- [x] Checked the items in the next section <sub>or this PR is labeled `deploy:gitlab`</sub>
- [x] PR is assigned to only the system administrator <sub>or this PR is not labeled `deploy:gitlab`</sub>

### System administrator

- [x] Background migrations for `dev.gitlab` are complete <sub>or this PR is not labeled `deploy:gitlab`</sub>
- [x] Background migrations for `anvildev.gitlab` are complete <sub>or this PR is not labeled `deploy:gitlab`</sub>
- [x] PR is assigned to only the operator

### Operator (before pushing merge the commit)

- [x] Ran `_select dev.gitlab && make -C terraform/gitlab/runner` <sub>or this PR is not labeled `deploy:runner`</sub>
- [x] Ran `_select anvildev.gitlab && make -C terraform/gitlab/runner` <sub>or this PR is not labeled `deploy:runner`</sub>
- [x] Added `sandbox` label <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed PR branch to GitLab `dev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed PR branch to GitLab `anvildev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `sandbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `anvilbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `sandbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `anvilbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Deleted unreferenced indices in `sandbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices in `dev`</sub>
- [x] Deleted unreferenced indices in `anvilbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices in `anvildev`</sub>
- [x] Started reindex in `sandbox` <sub>or this PR is not labeled `reindex:dev`</sub>
- [x] Started reindex in `anvilbox` <sub>or this PR is not labeled `reindex:anvildev`</sub>
- [x] Checked for failures in `sandbox` <sub>or this PR is not labeled `reindex:dev`</sub>
- [x] Checked for failures in `anvilbox` <sub>or this PR is not labeled `reindex:anvildev`</sub>
- [x] The title of the merge commit starts with the title of this PR
- [x] Added PR # reference to merge commit title
- [x] Collected commit title tags in merge commit title <sub>but only included `p` if the PR is also labeled `partial`</sub>
- [x] Moved connected issues to Merged column in ZenHub
- [x] Pushed merge commit to GitHub

### Operator (chain shortening)

- [x] Changed the target branch of the blocked PR to `develop` <sub>or this PR is not labeled `base`</sub>
- [x] Removed the `chained` label from the blocked PR <sub>or this PR is not labeled `base`</sub>
- [x] Removed the blocking relationship from the blocked PR <sub>or this PR is not labeled `base`</sub>
- [x] Removed the `base` label from this PR <sub>or this PR is not labeled `base`</sub>

### Operator (after pushing the merge commit)

- [x] Pushed merge commit to GitLab `dev`
- [x] Pushed merge commit to GitLab `anvildev`
- [x] Build passes on GitLab `dev`
- [x] Reviewed build logs for anomalies on GitLab `dev`
- [x] Build passes on GitLab `anvildev`
- [x] Reviewed build logs for anomalies on GitLab `anvildev`
- [x] Ran `_select dev.shared && make -C terraform/shared apply` <sub>or this PR is not labeled `deploy:shared`</sub>
- [x] Ran `_select anvildev.shared && make -C terraform/shared apply` <sub>or this PR is not labeled `deploy:shared`</sub>
- [x] Deleted PR branch from GitHub
- [x] Deleted PR branch from GitLab `dev`
- [x] Deleted PR branch from GitLab `anvildev`

### Operator (reindex)

- [x] Deindexed all unreferenced catalogs in `dev` <sub>or this PR is neither labeled `reindex:partial` nor `reindex:dev`</sub>
- [x] Deindexed all unreferenced catalogs in `anvildev` <sub>or this PR is neither labeled `reindex:partial` nor `reindex:anvildev`</sub>
- [x] Deindexed specific sources in `dev` <sub>or this PR is neither labeled `reindex:partial` nor `reindex:dev`</sub>
- [x] Deindexed specific sources in `anvildev` <sub>or this PR is neither labeled `reindex:partial` nor `reindex:anvildev`</sub>
- [x] Indexed specific sources in `dev` <sub>or this PR is neither labeled `reindex:partial` nor `reindex:dev`</sub>
- [x] Indexed specific sources in `anvildev` <sub>or this PR is neither labeled `reindex:partial` nor `reindex:anvildev`</sub>
- [x] Started reindex in `dev` <sub>or this PR does not require reindexing `dev`</sub>
- [x] Started reindex in `anvildev` <sub>or this PR does not require reindexing `anvildev`</sub>
- [x] Checked for, triaged and possibly requeued messages in both fail queues in `dev` <sub>or this PR does not require reindexing `dev`</sub>
- [x] Checked for, triaged and possibly requeued messages in both fail queues in `anvildev` <sub>or this PR does not require reindexing `anvildev`</sub>
- [x] Emptied fail queues in `dev` <sub>or this PR does not require reindexing `dev`</sub>
- [x] Emptied fail queues in `anvildev` <sub>or this PR does not require reindexing `anvildev`</sub>

### Operator

- [x] Propagated the `deploy:shared`, `deploy:gitlab`, `deploy:runner`, `reindex:partial`, `reindex:anvilprod` and `reindex:prod` labels to the next promotion PRs <sub>or this PR carries none of these labels</sub>
- [x] Propagated any specific instructions related to the `deploy:shared`, `deploy:gitlab`, `deploy:runner`, `reindex:partial`, `reindex:anvilprod` and `reindex:prod` labels, from the description of this PR to that of the next promotion PRs <sub>or this PR carries none of these labels</sub>
- [x] PR is assigned to no one

## Shorthand for review comments

- `L` line is too long
- `W` line wrapping is wrong
- `Q` bad quotes
- `F` other formatting problem

### CODE REVIEW

**Comment by nadove-ucsc** [comment]
Result of handover attempt: [workspace link](https://bvdp-saturn-dev.appspot.com/#workspaces/general-dev-billing-account/nadove-tmp-verbatim-nadove/data)

Screenshot from 2024-04-12 16-35-13

File: src/azul/service/avro_pfb.py [hunk 1]
```diff
@@ -674,54 +685,71 @@ def _update_replica_schema(*,
...
+        else:
+            if value is None and old_type == 'null':
                 pass
+            elif isinstance(value, list) and isinstance(old_type, dict) and old_type['type'] == 'array':
+                for v in value:
+                    _update_replica_schema_union(schema=old_type,
+                                                 path=path,
+                                                 key='items',
+                                                 value=v)
+            elif isinstance(value, dict) and isinstance(old_type, dict) and old_type['type'] == 'record':
```
Comment (achave11-ucsc) [inline src/azul/service/avro_pfb.py]: Maybe shortening it with something like this… 
```suggestion
            def old_type_is_of_type(type):
                return isinstance(old_type, dict) and old_type['type'] == type
            if value is None and old_type == 'null':
                pass
            elif isinstance(value, list) and old_type_is_of_type('array'):
                for v in value:
                    _update_replica_schema_union(schema=old_type,
                                                 path=path,
                                                 key='items',
                                                 value=v)
            elif isinstance(value, dict) and old_type_is_of_type('record'):
```
… but you may have a better solution.

File: src/azul/service/avro_pfb.py [hunk 2]
```diff
@@ -674,54 +685,71 @@ def _update_replica_schema(*,
...
-        if old_type == []:
-            schema[key] = _new_replica_schema(path=path, value=value)
-        elif value is None:
-            if old_type == 'null' or isinstance(old_type, list):
+        if isinstance(old_type, list):
+            _update_replica_schema_union(schema=schema, path=path, key=key, value=value)
+        else:
+            if value is None and old_type == 'null':
                 pass
+            elif isinstance(value, list) and isinstance(old_type, dict) and old_type['type'] == 'array':
```
Comment (achave11-ucsc) [inline src/azul/service/avro_pfb.py]: L

File: src/azul/service/avro_pfb.py (hunk 1 above)
Comment (achave11-ucsc) [inline src/azul/service/avro_pfb.py]: L

**Review by achave11-ucsc — CHANGES_REQUESTED** [review]
Excellent work, minor nits.

File: src/azul/service/avro_pfb.py (hunk 1 above)
Comment (nadove-ucsc) [reply src/azul/service/avro_pfb.py]: I prefer to just wrap the lines with parentheses rather than introduce a new function. This code is already hard to follow (and I wrote it :sweat_smile: ), so I figure it's best to keep things direct and inline, and it ends up taking less space once you add the obligatory blank line after the function body.

**Review by achave11-ucsc — DISMISSED** [review]
Approved ✅

File: test/service/test_manifest.py [hunk 3]
```diff
@@ -241,6 +241,30 @@ def sort_key(row: JSON) -> bytes:
         expected.sort(key=sort_key)
         self.assertEqual(expected, manifest)
 
+    def _assert_pfb(self,
```
Comment (hannes-ucsc) [inline test/service/test_manifest.py:242]: If we extract a method, we should simplify the contract between caller and callee. This method has side-effects on an argument, which is generally frowned upon, unless performance considerations make that necessary, in which case the docstring should make that clear. In this case, we should avoid the side-effects, either by not sorting, or at least not sorting in place.

File: test/service/test_manifest.py [hunk 4]
```diff
@@ -241,6 +241,30 @@ def sort_key(row: JSON) -> bytes:
...
+        :param expected_entities: a list of PFB entities.
+
+        :param actual: an HTTP response containing a PFB manifest.
+        """
+        manifest = fastavro.reader(BytesIO(actual.content))
+        schema = manifest.writer_schema
+        entities = list(manifest)
+        sort_key = compose_keys(none_safe_key(), itemgetter('id'))
+        entities.sort(key=sort_key)
+        expected_entities.sort(key=sort_key)
```
Comment (hannes-ucsc) [inline test/service/test_manifest.py]: Why do we need to do this? Can we sort the cans and commit them with the expected order?

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: test/service/test_manifest.py [hunk 5]
```diff
@@ -241,6 +241,29 @@ def sort_key(row: JSON) -> bytes:
         expected.sort(key=sort_key)
         self.assertEqual(expected, manifest)
 
+    def _assert_pfb(self,
+                    expected_schema: JSON,
+                    expected_entities: list[JSON],
```
Comment (hannes-ucsc) [inline test/service/test_manifest.py]: The type hints should reflect the contract wrt to mutability.

File: test/service/test_manifest.py [hunk 6]
```diff
@@ -241,6 +241,29 @@ def sort_key(row: JSON) -> bytes:
...
+                    actual: Response):
+        """
+        Assert that the body of the given response contains a valid PFB manifest
+        matching the expected schema and contents, disregarding the order of the
+        PFB entities.
```
Comment (hannes-ucsc) [inline test/service/test_manifest.py]: I may have been confused during my last review. *If* the verbatim manifests have non-deterministic ordering, there should be a comment that explains the reason for the non-determinism. I suspect that because the file-based manifest generators don't apply a pagination stage to the request, the hits are returned in an order that is not immediately apparent, but still deterministic. Ideally, there would be no sorting prior to the comparison, either of the actual value or the expected one. If we have to sort to work around a non-determinism, we should sort both the actual and the expected value prior to comparing them. Either way, the method should not have side-effects on its arguments and the type hints should reflect that.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
Don't need fixups next time.

File: test/service/test_manifest.py [hunk 7]
```diff
@@ -241,6 +242,29 @@ def sort_key(row: JSON) -> bytes:
...
+        :param expected_schema: a PFB schema.
+
+        :param expected_entities: a list of PFB entities.
+
+        :param actual: an HTTP response containing a PFB manifest.
+        """
+        manifest = fastavro.reader(BytesIO(actual.content))
+        schema = manifest.writer_schema
+        # The order of the entities in the manifest is deterministic, but varies
+        # depending on the current deployment, so there's no single ordering of
```
Comment (hannes-ucsc) [inline test/service/test_manifest.py]: This surprises me. Do you know why?

Comment (nadove-ucsc) [reply test/service/test_manifest.py]: When using the `scan` method without `preserve_order` set to True, hits are ordered by `_doc`, i.e., the order in which they are stored in the index. We observe that each shard is considered separately when determining this order, so that every document stored at position 0 in each shard is retrieved before each document stored at position 1 in each shard, and so on. 

The difference in ordering is caused by a difference in shard count, which is in turn caused by a difference in the (non-retry) contribution concurrency.

The number of ES shards for the `replicas` index is calculated [here](https://github.com/DataBiosphere/azul/blob/de54dcb6841362843f0b90cf9bd5a251155db1a6/src/azul/indexer/index_service.py#L163).

For personal deployments, where the contribution concurrency is only 8, the shard count is determined by the ES instance count, which is 2. For other lower deployments, the shard count is 8 (64 // 8) due to the higher concurrency. On the stable deployments, I would expect a shard count of 37 (300 // 8). 

The documents appear to be assigned to and retrieved from shards deterministically (there's no direct proof of this, but the ordering never changed in the course in my testing). Therefore, we should be able to stabilize the ordering by ensuring that the unit tests are always run with a consistent shard count.

Comment (nadove-ucsc) [reply test/service/test_manifest.py]: Ordering concerns postponed to https://github.com/DataBiosphere/azul/issues/6442

**Review by hannes-ucsc — APPROVED** [review]

**Comment by hannes-ucsc** [comment]
### Security design review

- [x] Security design review completed; this PR does **not** …
  - [x] … affect authentication; for example:
    - OAuth 2.0 with the application (API or Swagger UI)
    - Authentication of developers with Google Cloud APIs
    - Authentication of developers with AWS APIs
    - Authentication with a GitLab instance in the system
    - Password and 2FA authentication with GitHub
    - API access token authentication with GitHub
    - Authentication with Terra 
  - [x] … affect the permissions of internal users like access to
    - Cloud resources on AWS and GCP
    - GitLab repositories, projects and groups, administration
    - an EC2 instance via SSH
    - GitHub issues, pull requests, commits, commit statuses, wikis, repositories, organizations
  - [x] … affect the permissions of external users like access to
    - TDR snapshots
  - [x] … affect permissions of service or bot accounts
    - Cloud resources on AWS and GCP
  - [x] … affect audit logging in the system, like
    - adding, removing or changing a log message that represents an auditable event
    - changing the routing of log messages through the system
  - [x] … affect monitoring of the system
  - [x] … introduce a new software dependency like
    - Python packages on PYPI
    - Command-line utilities
    - Docker images
    - Terraform providers
  - [x] … add an interface that exposes sensitive or confidential data at the security boundary
  - [x] … affect the encryption of data at rest
  - [x] … require persistence of sensitive or confidential data that might require encryption at rest
  - [x] … require unencrypted transmission of data within the security boundary
  - [x] … affect the network security layer; for example by 
    - modifying, adding or removing firewall rules
    - modifying, adding or removing security groups
    - changing or adding a port a service, proxy or load balancer listens on
- [x] Documentation on any unchecked boxes is provided in comments below
